### ISSUE TITLE
PythonArgumentParser can't cope with `partialmethod`

### ISSUE THREAD

**Issue by dries007** [robotframework/robotframework#4318]
I'm currently writing some robot libraries where it would be quite handy to provide the user with a selection of variations of keywords with some arguments preset. To do this I thought I could use partialmethod, but unforutunately doing so makes the library unusable by RobotFramework 5.x.

Miniam code example:

```python
from functools import partialmethod


class PartialBug:
    def echo(self, x: str) -> str:
        return x

    hello_world = partialmethod(echo, x="hello world")
```

```robotframework
*** Settings ***
Library                 PartialBug
```

<details>
<summary>Stacktrace</summary>

```pytb
[ ERROR ] Unexpected error: AttributeError: 'functools.partial' object has no attribute '__name__'
Traceback (most recent call last):
  File "/home/dries/Projects/opso/opso-sw/.venv/lib/python3.10/site-packages/robot/utils/application.py", line 83, in _execute
    rc = self.main(arguments, **options)
  File "/home/dries/Projects/opso/opso-sw/.venv/lib/python3.10/site-packages/robot/run.py", line 450, in main
    result = suite.run(settings)
  File "/home/dries/Projects/opso/opso-sw/.venv/lib/python3.10/site-packages/robot/running/model.py", line 296, in run
    self.visit(runner)
  File "/home/dries/Projects/opso/opso-sw/.venv/lib/python3.10/site-packages/robot/model/testsuite.py", line 226, in visit
    visitor.visit_suite(self)
  File "/home/dries/Projects/opso/opso-sw/.venv/lib/python3.10/site-packages/robot/model/visitor.py", line 85, in visit_suite
    if self.start_suite(suite) is not False:
  File "/home/dries/Projects/opso/opso-sw/.venv/lib/python3.10/site-packages/robot/running/suiterunner.py", line 72, in start_suite
    ns.handle_imports()
  File "/home/dries/Projects/opso/opso-sw/.venv/lib/python3.10/site-packages/robot/running/namespace.py", line 55, in handle_imports
    self._handle_imports(self._imports)
  File "/home/dries/Projects/opso/opso-sw/.venv/lib/python3.10/site-packages/robot/running/namespace.py", line 66, in _handle_imports
    self._import(item)
  File "/home/dries/Projects/opso/opso-sw/.venv/lib/python3.10/site-packages/robot/running/namespace.py", line 74, in _import
    action(import_setting)
  File "/home/dries/Projects/opso/opso-sw/.venv/lib/python3.10/site-packages/robot/running/namespace.py", line 127, in _import_library
    lib = IMPORTER.import_library(name, import_setting.args,
  File "/home/dries/Projects/opso/opso-sw/.venv/lib/python3.10/site-packages/robot/running/importer.py", line 48, in import_library
    lib = self._import_library(name, positional, named, lib)
  File "/home/dries/Projects/opso/opso-sw/.venv/lib/python3.10/site-packages/robot/running/importer.py", line 78, in _import_library
    lib.create_handlers()
  File "/home/dries/Projects/opso/opso-sw/.venv/lib/python3.10/site-packages/robot/running/testlibraries.py", line 119, in create_handlers
    self._create_handlers(self.get_instance())
  File "/home/dries/Projects/opso/opso-sw/.venv/lib/python3.10/site-packages/robot/running/testlibraries.py", line 241, in _create_handlers
    handler, embedded = self._try_to_create_handler(name, method)
  File "/home/dries/Projects/opso/opso-sw/.venv/lib/python3.10/site-packages/robot/running/testlibraries.py", line 299, in _try_to_create_handler
    handler = self._create_handler(name, method)
  File "/home/dries/Projects/opso/opso-sw/.venv/lib/python3.10/site-packages/robot/running/testlibraries.py", line 310, in _create_handler
    return Handler(self, handler_name, handler_method)
  File "/home/dries/Projects/opso/opso-sw/.venv/lib/python3.10/site-packages/robot/running/handlers.py", line 40, in Handler
    return _PythonHandler(library, name, method)
  File "/home/dries/Projects/opso/opso-sw/.venv/lib/python3.10/site-packages/robot/running/handlers.py", line 143, in __init__
    _RunnableHandler.__init__(self, library, handler_name, handler_method,
  File "/home/dries/Projects/opso/opso-sw/.venv/lib/python3.10/site-packages/robot/running/handlers.py", line 60, in __init__
    self.arguments = self._parse_arguments(handler_method)
  File "/home/dries/Projects/opso/opso-sw/.venv/lib/python3.10/site-packages/robot/running/handlers.py", line 147, in _parse_arguments
    return PythonArgumentParser().parse(handler_method, self.longname)
  File "/home/dries/Projects/opso/opso-sw/.venv/lib/python3.10/site-packages/robot/running/arguments/py3argumentparser.py", line 29, in parse
    self._set_args(spec, handler)
  File "/home/dries/Projects/opso/opso-sw/.venv/lib/python3.10/site-packages/robot/running/arguments/py3argumentparser.py", line 43, in _set_args
    if handler.__name__ == '__init__':

Process finished with exit code 255
```

</details>

**Comment by pekkaklarck** [comment]
Thanks for the report. The provided example made it easy to reproduce the problem. It's not great that you cannot use `partialmethod`, but a much bigger problem is that the whole execution crashes. Luckily this is easy to fix.

I also tested creating a library using `partial`. There was no crash, but the keyword using `partial` didn't work. I haven't yet investigated this issue further.

The crash when using `partialmethod` is so bad that it definitely needs to be fixed already in RF 5.0.1. Probably best to add support for `partial` at the same time.

**Comment by pekkaklarck** [comment]
I investigated this a bit more. Turned out that the reason keywords created using `partial` are totally ignored is that [inspect.isroutine](https://python.readthedocs.io/en/stable/library/inspect.html#inspect.isroutine) we use internally returns false for them. I consider that odd and more like a bug than a feature. After a quick search on Python's issue tracker it seems others agree (https://github.com/python/cpython/issues/91002, https://github.com/python/cpython/issues/74315) so perhaps that's changed some day. If we want to support `partial` now, we need to somehow detect them separately.

Interestingly `isroutine` returns false also for methods created using `partialmethod` when accessing them from an instance (e.g. `isroutine(MyClass().method)` but true when accessing them directly from class (e.g. `isroutin(MyClass.method)`. We do the latter with class based libraries so thus keywords created using `partialmethod` aren't totally ignored similarly keywords created using `partial`. This then later causes the really bad crash because these partial methods don't have `__name__`.

We have three solutions:
1. Disallow using all kind of partial functions and methods. We basically just needed to handle the `partialmethod` case (they shouldn't become keywords in the first place) so this ought to be pretty easy. This isn't ideal because with later Python versions `isroutine` may behave differently and because not allowing partial methods as keywords is just a seemingly odd limitation.
2. Allow partial functions and methods. This requires handling the `partial` case (they should become keywords) and fixing the issue with missing `__name__`. The latter fix is rather trivial so this isn't a much bigger task than the above. 
3. Do option 1. now in RF 5.0.1 and option 2 in RF 5.1. Supporting partials is a new functionality and we don't generally add new features in patch releases like 5.0.1. I don't think being that pedantic to semantic versioning really benefits anyone, though, especially when this would be more work in total.

Regardless how this is handled, we need to add tests both for `partial` and `partialmethod`. That is especially important because their behavior may change in future Python versions.

**Comment by pekkaklarck** [comment]
I was able to fix this so that both `partial` and `partialmethod` works. It seems they expose argument information of the underlying function/method so that Robot's argument validation and conversion works fine. Keyword docs won't be that useful, though.

**Comment by dries007** [comment]
Thanks! I presume that means it will be included in the next release?

**Comment by pekkaklarck** [comment]
Yeah, the fix will be in 5.0.1. Adding support for partial objects would have otherwise belonged to 5.1, but the crash needed to be fixed anyway.
